external/fv3fit/fv3fit/pytorch/graph/mpg_unet.py

```diff
@@ -0,0 +1,331 @@
+        x1 = x1.reshape(
+            x1.size(0) * x1.size(1), x1.size(2), x1.size(3), x1.size(4)
+        )  # change the shape to (batch_size*n_tiles, n_features, n_x, n_y )
+        x1 = self.up(x1)
+        x1 = x1.reshape(batch_size, n_tiles, n_features // 2, n_x * 2, n_y * 2,)
+        x1 = x1.permute(
+            0, 1, 3, 4, 2
+        )  # change dimensions to (batch_size, n_tiles, n_x, n_y, n_features)
+        return x1
+
+
+class UNet(nn.Module):
+    """
+    A graph based U-net architucture
+    """
+
+    def __init__(
+        self, config, down_factory, up_factory, depth: int, in_channels: int, nx: int
+    ):
+        """
+        Args:
+            down_factory: double-convolution followed
+                by a pooling layer on the down-path side
+            up_factory: Upsampling followed by
+                double-convolution on the up-path side
+            depth: depth of the UNet
+            in_channels: number of input channels
+        """
+        super(UNet, self).__init__()
+
+        node_lower_channels = 2 * in_channels
+        self._down = down_factory()
+
+        self.conv1 = MPNNGNN(
+            in_channels,
+            node_lower_channels,
+            config.edge_hidden_features,
+            config.num_step_message_passing,
+            config.activation,
+            config.aggregator,
+            nx,
+        )
+
+        self.conv2 = MPNNGNN(
+            2 * node_lower_channels,
+            node_lower_channels,
+            config.edge_hidden_features,
+            config.num_step_message_passing,
+            config.activation,
+            config.aggregator,
+            nx,
+        )
+
+        if depth == 1:
+            self._lower = MPNNGNN(
+                node_lower_channels,
+                node_lower_channels * 2,
+                config.edge_hidden_features,
+                config.num_step_message_passing,
+                config.activation,
+                config.aggregator,
+                nx=nx // 2,
+            )
+
+        elif depth <= 0:
+            raise ValueError(f"depth must be at least 1, got {depth}")
+        else:
+            self._lower = UNet(
+                config,
+                down_factory,
+                up_factory,
+                depth=depth - 1,
+                in_channels=node_lower_channels,
+                nx=nx // 2,
+            )
+        self._up = up_factory(in_channels=node_lower_channels * 2)
+
+    def forward(self, inputs):
+        before_pooling = self.conv1(inputs)
+        x = self._down(before_pooling)
+        x = self._lower(x)
+        x = self._up(x)
+        x = torch.cat([before_pooling, x], dim=-1)
+        x = self.conv2(x)
+        return x
+
+
+class MPGraphUNet(nn.Module):
+    def __init__(self, config, in_channels: int, out_channels: int, nx: int):
+        """
+        Args:
+            in_channels: number of input channels
+        """
```

Comment:
Docstring does not match init method.

Please add type hints to all `config` arguments.